File: SeparatingAxisTest/Range.m
//
//  Range.m
//  SeparatingAxisTest
//
//  Created by Najati Imam on 7/26/12.
//  Copyright (c) 2012 Cyrus Innovation. All rights reserved.
//

#import "Range.h"

@implementation Range

@synthesize min;
@synthesize max;

-(id) initWithMax:(float) _max andMin:(float) _min {
    if (self = [super init]) {
        max = _max;
        min = _min;
    }
    return self;
}

-(Overlap *) overlapWith:(Range *) that {
    Boolean overlaps = !(([self max] < [that min]) || ([that max] < [self min]));
    
    if (!overlaps) {
        return [[Overlap alloc] initFrom:false and:0];
    }
    
    float left = -([self max] - [that min]);
    float right = [that max] - [self min];
    float correction = fabsf(left) < fabsf(right) ? left : right;
    return [[Overlap alloc] initFrom:true and:correction];
}


@end

File: iOSGeometryTools/Range.c
//
//  Range.m
//  SeparatingAxisTest
//
//  Created by Najati Imam on 7/26/12.
//  Copyright (c) 2012 Cyrus Innovation. All rights reserved.
//

#include "Range.h"

#include <stdbool.h>
#include <math.h>

Range range_from(float min, float max) {
    Range r;
    r.min = min;
    r.max = max;
    return r;
}

Overlap create_overlap(Range a, Range b) {
    bool overlaps = !((a.max < b.min) || (b.max < a.min));
    
    Overlap overlap;
    
    if (!overlaps) {
        overlap.overlaps = false;
        overlap.correction = 0;
    } else {
        float left = -(a.max - b.min);
        float right = b.max - a.min;
        float correction = fabsf(left) < fabsf(right) ? left : right;
        
        overlap.overlaps = true;
        overlap.correction = correction;
    }
    
    return overlap;
}
